content/blog/2019-12-03-posits/index.md

```diff
@@ -0,0 +1,262 @@
+- `convert typ to double`: conversion from `typ` to `double`
+- `typ_op`: implementation over `typ` values of `op`, the corresponding `double` operation
+
+The `float` pass specifies these components as:
+- `fptrunc double to float`
+- `fpext float to double`
+- LLVM's floating point operations with type parameter set to `float` (e.g., `fadd float x y`)
+
+The `posit` pass draws these components from external `C` functions implemented in [Cerlane Leong's SoftPosit repository](https://gitlab.com/cerlane/SoftPosit-Python).
+In particular, we borrow the basic arithmetic operations (`+`, `-`, `*`, and `/`) over 32-bit posits. 
+
+
+### Benchmark Results
+
+We ran several benchmarks from [FPBench](https://fpbench.org/benchmarks.html) that accomodated our limited selection of posit operators.
+Here, we list their names, sample sizes, and associated results:
+
+
+| `sum` | Mean Error | Error Std Dev |
+|-------|------------|---------------|
+| Float | 7.7e-8     | 4.8e-8        |
+| Posit | 5.1e-9     | 3.2e-9        |
+|`n`= 1000 |||
+
+
+| `x_by_xy` | Mean Error | Error Std Dev |
+|-------|------------|---------------|
+| Float | 1.1e-7     | 9.2e-8        |
+| Posit | 6.9e-9     | 5.7e-9        |
+| `n` = 961 | | |
+
+
+| `delta4` | Mean Error | Error Std Dev |
+|-------|------------|---------------|
+| Float | 5.8e-7     | 6.5e-6        |
+| Posit | 3.6e-8     | 3.8e-7        |
+| `n`= 262144 |   |    |
+
+
+
+| `kepler0` | Mean Error | Error Std Dev |
+|-------|------------|---------------|
+| Float | 2.5e-7     | 1.0e-7        |
+| Posit | 1.6e-8     | 7.9e-9        |
+| `n`= 262144  |  |        |
+
+
+| `floudas1`  | Mean Error | Error Std Dev |
+|-------|------------|---------------|
+| Float | 2.1e-7     | 1.6e-7        |
+| Posit | 1.4e-8     | 1.2e-8        |
+| `n`= 92572956 |   |        |
+
+
+# Discussion
+
+As prophesied by Gustafson, posits consistently outperform floating points by orders of magnitude.
+However, our LLVM `posit` pass is grossly inefficient: it introduces loads and stores operations to convert between `double` and `posit` for each `double` operation.
+We blame this sad state of affairs on ignorant, non-posit architectures, which necessitate such a pass in the first place.
+
+Gustafson shares what we're missing out on by not being woke:
+>  A posit processing unit takes less circuitry than an IEEE float FPU. With lower power use and smaller silicon footprint, the posit operations per second (POPS) supported by a chip can be significantly higher than the FLOPS using similar hardware resources. GPU accelerators and Deep Learning processors, in particular,can do more per watt and per dollar with posits, yet deliver superior answer quality
+Indeed, we could be talking in terms of POPS instead of FLOPS.
+Nonetheless, our pass allows us to simulate and give thanks for posit accuracy.
+
+As mentioned previously, our approach treats `double` as the "ground truth" representation for benchmarks. 
```

Comment:
But perhaps it bears mentioning one more time. :)

Reply:
Do you mean that the fact of the "previous mentioning" is superfluous? I assume so, so I removed that transition.